File: src/Orison.Core/Components/Transform3D.cs
namespace Orison {
    /// <summary>
    /// 3D transform component for positioning, rotating, and scaling entities in 3D space.
    /// </summary>
    public class Transform3D : Component {
        private Vector3 position;
        private Vector3 rotation; // Euler angles in radians
        private Vector3 scale;

        /// <summary>
        /// The position in 3D space.
        /// </summary>
        public Vector3 Position {
            get => position;
            set => position = value;
        }

        /// <summary>
        /// The rotation as Euler angles (radians).
        /// </summary>
        public Vector3 Rotation {
            get => rotation;
            set => rotation = value;
        }

        /// <summary>
        /// The scale.
        /// </summary>
        public Vector3 Scale {
            get => scale;
            set => scale = value;
        }

        public Transform3D() {
            position = Vector3.Zero;
            rotation = Vector3.Zero;
            scale = Vector3.One;
        }

        public Transform3D(Vector3 position) {
            this.position = position;
            this.rotation = Vector3.Zero;
            this.scale = Vector3.One;
        }

        public Transform3D(Vector3 position, Vector3 rotation) {
            this.position = position;
            this.rotation = rotation;
            this.scale = Vector3.One;
        }

        public Transform3D(Vector3 position, Vector3 rotation, Vector3 scale) {
            this.position = position;
            this.rotation = rotation;
            this.scale = scale;
        }

        /// <summary>
        /// Set the position.
        /// </summary>
        public void SetPosition(float x, float y, float z) {
            position = new Vector3(x, y, z);
        }

        /// <summary>
        /// Set the rotation (in radians).
        /// </summary>
        public void SetRotation(float x, float y, float z) {
            rotation = new Vector3(x, y, z);
        }

        /// <summary>
        /// Set the scale.
        /// </summary>
        public void SetScale(float x, float y, float z) {
            scale = new Vector3(x, y, z);
        }

        /// <summary>
        /// Translate the position.
        /// </summary>
        public void Translate(float x, float y, float z) {
            position += new Vector3(x, y, z);
        }

        /// <summary>
        /// Translate by a vector.
        /// </summary>
        public void Translate(Vector3 delta) {
            position += delta;
        }

        /// <summary>
        /// Rotate the object.
        /// </summary>
        public void Rotate(float x, float y, float z) {
            rotation += new Vector3(x, y, z);
        }

        /// <summary>
        /// Rotate by a vector.
        /// </summary>
        public void Rotate(Vector3 delta) {
            rotation += delta;
        }

        /// <summary>
        /// Scale the object.
        /// </summary>
        public void ScaleBy(float x, float y, float z) {
            scale *= new Vector3(x, y, z);
        }

        /// <summary>
        /// Scale by a vector.
        /// </summary>
        public void ScaleBy(Vector3 factor) {
            scale *= factor;
        }

        /// <summary>
        /// Get the forward direction vector.
        /// </summary>
        public Vector3 Forward() {
            // Simple forward calculation based on Y rotation
            float cosY = (float)Math.Cos(rotation.Y);
            float sinY = (float)Math.Sin(rotation.Y);
            return new Vector3(-sinY, 0, -cosY);
        }

        /// <summary>
        /// Get the right direction vector.
        /// </summary>
        public Vector3 Right() {
            // Simple right calculation based on Y rotation
            float cosY = (float)Math.Cos(rotation.Y);
            float sinY = (float)Math.Sin(rotation.Y);
            return new Vector3(cosY, 0, -sinY);
        }

        /// <summary>
        /// Get the up direction vector.
        /// </summary>
        public Vector3 Up() {
            return Vector3.Up;
        }
    }
}

File: samples/orison-web/src/components/Transform3D.ts
import { Component } from '../core/Component';
import * as THREE from 'three';

/**
 * 3D transform component for positioning, rotating, and scaling entities in 3D space.
 */
export class Transform3D extends Component {
  private position: THREE.Vector3;
  private rotation: THREE.Euler;
  private scale: THREE.Vector3;
  private threeObject: THREE.Object3D;

  constructor() {
    super();
    this.position = new THREE.Vector3(0, 0, 0);
    this.rotation = new THREE.Euler(0, 0, 0);
    this.scale = new THREE.Vector3(1, 1, 1);
    this.threeObject = new THREE.Object3D();
  }

  /**
   * Gets the position.
   */
  get Position(): THREE.Vector3 {
    return this.position;
  }

  /**
   * Gets the rotation.
   */
  get Rotation(): THREE.Euler {
    return this.rotation;
  }

  /**
   * Gets the scale.
   */
  get Scale(): THREE.Vector3 {
    return this.scale;
  }

  /**
   * Gets the Three.js Object3D.
   */
  get ThreeObject(): THREE.Object3D {
    return this.threeObject;
  }

  /**
   * Sets the position.
   */
  setPosition(x: number, y: number, z: number): void {
    this.position.set(x, y, z);
    this.threeObject.position.set(x, y, z);
  }

  /**
   * Sets the rotation (in radians).
   */
  setRotation(x: number, y: number, z: number): void {
    this.rotation.set(x, y, z);
    this.threeObject.rotation.set(x, y, z);
  }

  /**
   * Sets the scale.
   */
  setScale(x: number, y: number, z: number): void {
    this.scale.set(x, y, z);
    this.threeObject.scale.set(x, y, z);
  }

  /**
   * Translates the position.
   */
  translate(x: number, y: number, z: number): void {
    this.position.add(new THREE.Vector3(x, y, z));
    this.threeObject.position.add(new THREE.Vector3(x, y, z));
  }

  /**
   * Rotates the object.
   */
  rotate(x: number, y: number, z: number): void {
    this.rotation.x += x;
    this.rotation.y += y;
    this.rotation.z += z;
    this.threeObject.rotation.x += x;
    this.threeObject.rotation.y += y;
    this.threeObject.rotation.z += z;
  }

  /**
   * Scales the object.
   */
  scaleBy(x: number, y: number, z: number): void {
    this.scale.multiply(new THREE.Vector3(x, y, z));
    this.threeObject.scale.multiply(new THREE.Vector3(x, y, z));
  }

  /**
   * Looks at a target position.
   */
  lookAt(target: THREE.Vector3): void {
    this.threeObject.lookAt(target);
  }

  override update(deltaTime: number): void {
    // Sync Three.js object with transform values
    this.threeObject.position.copy(this.position);
    this.threeObject.rotation.copy(this.rotation);
    this.threeObject.scale.copy(this.scale);
  }
}
